Bug 128772 - Account Help Button does not work

Bob-IT opened the pull request:
I am unsure about this one, I was looking at the bug and the context help was not working. I do not know how long it has not been working but hopefully not from the report date.

From testing, when pressing the help button, the following uri is created for the new account help button...
'ghelp:gnucash-help?acct-create' and used to open the gnome help app.

This segfaults in yelp at the '?' as 'ghelp:gnucash-help' does work.
To prove try at a command prompt...
gnome-help ghelp:gnucash-help?acct-create which segfaults but this...
gnome-help ghelp:/usr/share/gnome/help/gnucash-help/C/gnucash-help.xml?acct-create works.

To fix this, I have changed the context help to use the full path to the help files.
Looking at this I think I should add a test for getting a valid path and maybe an error dialog but will wait for any feedback first.

jralls commented:
Please split that into a separate static function.

I'm a little surprised that there have been so few complaints about this. Makes one suspect that context help just isn't used very much and that maintaining it is wasted effort.

fellen commented on gnucash/gnome-utils/gnc-gnome-utils.c:
```diff
@@ -421,7 +421,39 @@ gnc_gnome_help (const char *file_name, const char *anchor)
+            g_free (filename);
+        }
+
+        help_file = g_strconcat (file_name, ".xml", NULL);
```
".xml" should also get a more prominent place. The documenters could decide to use ".dokbook" or switch to another format.

jralls replied:
Yelp can handle .html, maybe we should switch to that (not in this PR of course). Even better, we could ditch yelp and Windows help viewer entirely and display the docs with gnc-html.

fellen commented on gnucash/gnome-utils/gnc-gnome-utils.c:
```diff
@@ -421,7 +421,39 @@ gnc_gnome_help (const char *file_name, const char *anchor)
+        }
+
+        if (out)
+            uri = g_strconcat ("ghelp:", out, "?", anchor, NULL);
```
How about making it smarter. Imagine anchor was added to C, but de got no update. Can we then fall back to C?

Bob-IT replied:
Can not see an easy way of doing this.

fellen commented:
> I'm a little surprised that there have been so few complaints about this. Makes one suspect that context help just isn't used very much and that maintaining it is wasted effort.

I still remember, when it occured to me first, I didn't know about bugzilla, mailinglists - probably not in my language -, …
I simply thought "Another low quality program".

IMHO it is a bootstrapping problem: If  I can not start the docs, I do not know, how to complain.

jralls commented:
> IMHO it is a bootstrapping problem: If I can not start the docs, I do not know, how to complain.

This isn't about not being able to start the docs, it's about context help and even when that works it won't tell you anything about filing bugs. That information lives in the T&CG.

Speaking of bugs, [one of yours](https://bugs.gnucash.org/show_bug.cgi?id=796852) indicates that the low-quality program would be yelp. IMO dropping support for that and the seriously obsolete Windows help viewer is a better way forward. We already have the infrastructure to display html in GnuCash.

gjanssens commented:
The bug is in yelp really, not in gnucash. According to their own user manual our incantation should be supported:
https://people.gnome.org/~bmsmith/build/yelp.html#yelp-advanced-cmdline

So you could work around it in GnuCash as a service to our users, but better would be to fix it in yelp (ie file a  bug at yelp or provide a patch to their code to get it fixed). This is how the open source ecosystem as a whole advances after all.

Having said all that though, I do agree with @jralls it's probably better to just switch to using our built-in webkit to display help.

fellen commented:
> The bug is in yelp really, not in gnucash. According to their own user manual our incantation should be supported:
> https://people.gnome.org/~bmsmith/build/yelp.html#yelp-advanced-cmdline

That is the 14years old Gnome2 variant - with scolkeeper, …
It's a shame that documentation tools are so badly documented. ;-) 
 > So you could work around it in GnuCash as a service to our users, but better would be to fix it in yelp (ie file a bug at yelp

https://gitlab.gnome.org/GNOME/yelp/issues/116
> or provide a patch to their code to get it fixed). This is how the open source ecosystem as a whole advances after all.
> 
> Having said all that though, I do agree with @jralls it's probably better to just switch to using our built-in webkit to display help.

There have been more other options been discussed on the bugs, IIRC.

gjanssens commented:
Thanks for the link to the yelp bug @fellen 
Looks like we shouldn't wait for that to get fixed :(

gjanssens commented:
Also, while I'm fine with this workaround, I would recommend adding a comment on top explaining why it is necessary. If for some reason switching to our built-in webkit takes a long time, we may forget and try to simplify it again at some point.

Bob-IT commented:
I have created a separate function and added testing for a valid path and reporting error with the use of gnc_error_dialog. 
Also added comment for why it is there with reference to yelp bug report.
As a result of testing, I have added an other commit to add a GtkWindow parameter to gnc_gnome_help so the parent window can be specified for the error dialog which resulted in the last commit for the commodity help. Could not see why it was setup that way so changed it to be like others.

Bob-IT commented:
I have pushed this as a temporary fix and will have a look at creating a help dialog using existing 'gnc-html'

fellen commented:
> I have pushed this as a temporary fix and will have a look at creating a help dialog using existing 'gnc-html'

Another approach would be to talk with libyelp directly. But I am not sure, it is it worth.
